File: src/layout_actions.py
"""Process-preserving pane reshape and resize.

Adapted from herdr-pane-layouts `src/cli.py` (commit
0ef0a8d5d463757f06037551fc1f5ef6dfde478d). Mutations run under ctx.Lock and
talk to Herdr through rpc.py.
"""

import os
import sys

import ctx
import layouts
import rpc


RESIZE_AMOUNT = 0.02
RESIZE_DIRECTIONS = {
    "resize-left": "left",
    "resize-down": "down",
    "resize-up": "up",
    "resize-right": "right",
}


class LayoutError(RuntimeError):
    pass


def _notify(message):
    rpc.try_call(
        "notification.show",
        {"title": "Mosaic layouts failed", "body": message, "sound": "none"},
    )


def _pane_context():
    pane_id = os.environ.get("HERDR_PANE_ID")
    if pane_id:
        return pane_id
    c = ctx.invocation_context()
    return c.get("focused_pane_id") or c.get("pane_id")


def _export_layout(tab_id=None):
    pane_id = None if tab_id else _pane_context()
    params = {}
    if tab_id:
        params["tab_id"] = tab_id
    elif pane_id:
        params["pane_id"] = pane_id
    result = rpc.call("layout.export", params)
    layout = (result or {}).get("layout")
    if not layout:
        raise LayoutError("layout.export returned no layout")
    return layout


def _tab_destination(tab_id, target_pane_id, direction, ratio):
    return {
        "type": "tab",
        "tab_id": tab_id,
        "target_pane_id": target_pane_id,
        "split": direction,
        "ratio": ratio,
    }


def _move(pane_id, destination, focus=False):
    result = rpc.call(
        "pane.move",
        {"pane_id": pane_id, "destination": destination, "focus": focus},
    )
    move_result = (result or {}).get("move_result") or {}
    if not move_result.get("changed"):
        raise LayoutError("pane.move: %s" % (move_result.get("reason") or "move rejected"))
    return move_result


def recover(staging_tab, original_tab, focused_pane):
    staged = layouts.pane_ids(_export_layout(tab_id=staging_tab)["root"])
    target = layouts.first_pane(_export_layout(tab_id=original_tab)["root"])
    for pane_id in staged:
        _move(
            pane_id,
            _tab_destination(original_tab, target, "right", 0.5),
            focus=pane_id == focused_pane,
        )


def reshape(layout, target):
    ids = layouts.pane_ids(layout["root"])
    if len(ids) < 2 or layouts.same(layout["root"], target):
        return
    if layout.get("zoomed"):
        raise LayoutError("unzoom the tab before changing its layout")

    tab_id = layout["tab_id"]
    focused = layout["focused_pane_id"]
    anchor = layouts.first_pane(target)
    staged = [pane_id for pane_id in ids if pane_id != anchor]
    current = dict((pane_id, pane_id) for pane_id in ids)
    staging_tab = None

    try:
        first = staged[0]
        moved = _move(
            first,
            {
                "type": "new_tab",
                "workspace_id": layout["workspace_id"],
                "label": "layout-staging-%d" % os.getpid(),
            },
        )
        created = moved.get("created_tab") or {}
        staging_tab = created.get("tab_id")
        if not staging_tab:
            raise LayoutError("pane.move did not create a staging tab")
        current[first] = moved["pane"]["pane_id"]
        staging_target = current[first]

        for pane_id in staged[1:]:
            moved = _move(
                current[pane_id],
                _tab_destination(staging_tab, staging_target, "right", 0.5),
            )
            current[pane_id] = moved["pane"]["pane_id"]

        for target_id, source_id, direction, ratio in layouts.insertion_plan(target):
            moved = _move(
                current[source_id],
                _tab_destination(tab_id, current[target_id], direction, ratio),
                focus=source_id == focused,
            )
            current[source_id] = moved["pane"]["pane_id"]

        staging_tab = None
    except Exception:
        if staging_tab:
            try:
                recover(staging_tab, tab_id, current[focused])
            except Exception as error:
                sys.stderr.write(
                    "mosaic: recovery failed; panes remain in %s: %s\n"
                    % (staging_tab, error)
                )
        raise


def target_for(action, layout):
    choices = layouts.presets(layouts.pane_ids(layout["root"]))
    if action == "equalize":
        return choices[0][1]
    current = -1
    for index, (_name, tree) in enumerate(choices):
        if layouts.same(layout["root"], tree):
            current = index
            break
    return choices[(current + 1) % len(choices)][1]


def run_action(action):
    """Execute one layout action. Caller holds ctx.Lock."""
    direction = RESIZE_DIRECTIONS.get(action)
    if direction:
        pane_id = _pane_context()
        if not pane_id:
            raise LayoutError("resize action requires a pane context")
        rpc.call(
            "pane.resize",
            {"pane_id": pane_id, "direction": direction, "amount": RESIZE_AMOUNT},
        )
        return 0
    if action in ("equalize", "cycle"):
        current = _export_layout()
        reshape(current, target_for(action, current))
        return 0
    raise LayoutError("unknown layout action %r" % action)


def run(argv):
    action = argv[0] if argv else ""
    try:
        with ctx.Lock():
            return run_action(action)
    except (LayoutError, rpc.RpcError, OSError, KeyError, ValueError) as error:
        ctx.warn("layouts: %s" % error)
        _notify(str(error))
        print("mosaic: %s" % error, file=sys.stderr)
        return 1

File: src/panes/layout-actions.ts
import { Effect, Predicate, Result, Schema } from "effect"

import { pluginLockPath, withExclusiveLock } from "../runtime/lock.ts"
import { FlockError, LayoutError, RpcTransportError } from "../runtime/errors.ts"
import { paneIdFromContextJson } from "../runtime/invocation.ts"
import {
  emptyOutput,
  joinOutput,
  pluginWarn,
  type CapturedOutput,
} from "../runtime/plugin-log.ts"
import type { PluginPathValues } from "../runtime/paths.ts"
import { PluginPaths } from "../runtime/paths.ts"
import { rpcCall, rpcTryCall, type JsonObject } from "../runtime/rpc.ts"
import {
  firstPane,
  insertionPlan,
  paneIds,
  same,
  targetFor,
  type LayoutNode,
} from "./layouts.ts"

type Json = typeof Schema.Json.Type

export const RESIZE_AMOUNT = 0.02

const RESIZE_DIRECTIONS = {
  "resize-left": "left",
  "resize-down": "down",
  "resize-up": "up",
  "resize-right": "right",
} as const

export type ExportedLayout = {
  readonly tab_id: string
  readonly workspace_id: string
  readonly focused_pane_id: string
  readonly zoomed: boolean
  readonly root: LayoutNode
}

function commandResult(code: number, output: CapturedOutput) {
  return {
    code,
    stdout: joinOutput(output.stdout),
    stderr: joinOutput(output.stderr),
  } as const
}

function pythonRepr(value: string): string {
  return `'${value.replace(/\\/g, "\\\\").replace(/'/g, "\\'")}'`
}

function asObject(value: Json | undefined): JsonObject | undefined {
  const decoded = Schema.decodeUnknownResult(Schema.JsonObject)(value ?? null)

  if (Result.isFailure(decoded)) return undefined

  return decoded.success
}

function asJsonObject(value: Json): JsonObject {
  const decoded = Schema.decodeUnknownResult(Schema.JsonObject)(value)

  if (Result.isFailure(decoded)) return {}

  return decoded.success
}

function paneContext(paths: PluginPathValues): string | undefined {
  if (paths.paneId !== undefined && paths.paneId !== "") return paths.paneId

  return paneIdFromContextJson(paths.contextJson)
}

function errorText(error: LayoutError | RpcTransportError | FlockError | Error): string {
  if (error instanceof LayoutError) return error.message

  if (error instanceof RpcTransportError) return `${error.code}: ${error.message}`

  return error.message
}

function parseNode(value: Json | undefined): LayoutNode {
  const object = asObject(value)

  if (object === undefined) {
    throw new LayoutError({ message: "layout.export returned no layout" })
  }

  if (object.type === "pane") {
    const paneId = object.pane_id

    if (!Predicate.isString(paneId) || paneId === "") {
      throw new LayoutError({ message: "layout contains a pane without an id" })
    }

    return { type: "pane", pane_id: paneId }
  }

  if (object.type === "split") {
    const direction = object.direction
    const ratio = object.ratio

    if (!Predicate.isString(direction) || !Predicate.isNumber(ratio)) {
      throw new LayoutError({ message: "layout.export returned no layout" })
    }

    return {
      type: "split",
      direction,
      ratio,
      first: parseNode(object.first),
      second: parseNode(object.second),
    }
  }

  throw new LayoutError({ message: "layout.export returned no layout" })
}

function parseExportedLayout(value: JsonObject): ExportedLayout {
  const tabId = value.tab_id
  const workspaceId = value.workspace_id
  const focused = value.focused_pane_id

  if (!Predicate.isString(tabId) || tabId === "") {
    throw new LayoutError({ message: "layout.export returned no layout" })
  }

  if (!Predicate.isString(workspaceId) || workspaceId === "") {
    throw new LayoutError({ message: "layout.export returned no layout" })
  }

  if (!Predicate.isString(focused) || focused === "") {
    throw new LayoutError({ message: "layout.export returned no layout" })
  }

  return {
    tab_id: tabId,
    workspace_id: workspaceId,
    focused_pane_id: focused,
    zoomed: value.zoomed === true,
    root: parseNode(value.root),
  }
}

const notifyFailure = Effect.fnUntraced(function*(message: string) {
  yield* rpcTryCall("notification.show", {
    title: "Mosaic layouts failed",
    body: message,
    sound: "none",
  })
})

const exportLayout = Effect.fnUntraced(function*(
  paths: PluginPathValues,
  tabId: string | undefined,
) {
  const paneId = tabId === undefined ? paneContext(paths) : undefined

  const payload = yield* rpcCall(
    "layout.export",
    tabId !== undefined
      ? asJsonObject({ tab_id: tabId })
      : paneId !== undefined
        ? asJsonObject({ pane_id: paneId })
        : {},
  )

  const layout = asObject(payload.layout)

  if (layout === undefined) {
    return yield* new LayoutError({ message: "layout.export returned no layout" })
  }

  try {
    return parseExportedLayout(layout)
  } catch (error) {
    if (error instanceof LayoutError) return yield* error

    if (error instanceof Error) {
      return yield* new LayoutError({ message: error.message })
    }

    return yield* new LayoutError({ message: "layout.export returned no layout" })
  }
})

function tabDestination(
  tabId: string,
  targetPaneId: string,
  direction: string,
  ratio: number,
): JsonObject {
  return asJsonObject({
    type: "tab",
    tab_id: tabId,
    target_pane_id: targetPaneId,
    split: direction,
    ratio,
  })
}

function newTabDestination(workspaceId: string, label: string): JsonObject {
  return asJsonObject({
    type: "new_tab",
    workspace_id: workspaceId,
    label,
  })
}

const movePane = Effect.fnUntraced(function*(
  paneId: string,
  destination: JsonObject,
  focus = false,
) {
  const payload = yield* rpcCall("pane.move", {
    pane_id: paneId,
    destination,
    focus,
  })

  const moveResult = asObject(payload.move_result) ?? {}

  if (moveResult.changed !== true) {
    const reason = Predicate.isString(moveResult.reason) && moveResult.reason !== ""
      ? moveResult.reason
      : "move rejected"

    return yield* new LayoutError({ message: `pane.move: ${reason}` })
  }

  return moveResult
})

function movedPaneId(moveResult: JsonObject, fallback: string): string {
  const pane = asObject(moveResult.pane)
  const paneId = pane === undefined ? undefined : pane.pane_id

  return Predicate.isString(paneId) && paneId !== "" ? paneId : fallback
}

function createdTabId(moveResult: JsonObject): string | undefined {
  const created = asObject(moveResult.created_tab)
  const tabId = created === undefined ? undefined : created.tab_id

  return Predicate.isString(tabId) && tabId !== "" ? tabId : undefined
}

const recover = Effect.fnUntraced(function*(
  paths: PluginPathValues,
  stagingTab: string,
  originalTab: string,
  focusedPane: string,
) {
  const staged = paneIds((yield* exportLayout(paths, stagingTab)).root)
  const target = firstPane((yield* exportLayout(paths, originalTab)).root)

  for (const paneId of staged) {
    yield* movePane(
      paneId,
      tabDestination(originalTab, target, "right", 0.5),
      paneId === focusedPane,
    )
  }
})

const rearrangePanes = Effect.fnUntraced(function*(
  paths: PluginPathValues,
  output: CapturedOutput,
  layout: ExportedLayout,
  target: LayoutNode,
) {
  const ids = paneIds(layout.root)

  if (ids.length < 2 || same(layout.root, target)) return

  if (layout.zoomed) {
    return yield* new LayoutError({ message: "unzoom the tab before changing its layout" })
  }

  const tabId = layout.tab_id
  const focused = layout.focused_pane_id
  const anchor = firstPane(target)
  const staged = ids.filter((paneId) => paneId !== anchor)
  const current = new Map<string, string>()

  for (const paneId of ids) current.set(paneId, paneId)

  const firstStaged = staged[0]

  if (firstStaged === undefined) return

  let stagingTab: string | undefined

  const applied = yield* Effect.gen(function*() {
    const firstMoved = yield* movePane(
      firstStaged,
      newTabDestination(layout.workspace_id, `layout-staging-${process.pid}`),
    )

    const created = createdTabId(firstMoved)

    if (created === undefined) {
      return yield* new LayoutError({ message: "pane.move did not create a staging tab" })
    }

    stagingTab = created
    current.set(firstStaged, movedPaneId(firstMoved, firstStaged))
    const stagingTarget = current.get(firstStaged) ?? firstStaged

    for (const paneId of staged.slice(1)) {
      const liveId = current.get(paneId) ?? paneId

      const moved = yield* movePane(
        liveId,
        tabDestination(created, stagingTarget, "right", 0.5),
      )

      current.set(paneId, movedPaneId(moved, liveId))
    }

    for (const step of insertionPlan(target)) {
      const sourceLive = current.get(step.sourceId) ?? step.sourceId
      const targetLive = current.get(step.targetId) ?? step.targetId

      const moved = yield* movePane(
        sourceLive,
        tabDestination(tabId, targetLive, step.direction, step.ratio),
        step.sourceId === focused,
      )

      current.set(step.sourceId, movedPaneId(moved, sourceLive))
    }

    stagingTab = undefined
  }).pipe(Effect.result)

  if (Result.isSuccess(applied)) return

  if (stagingTab !== undefined) {
    const focusedLive = current.get(focused) ?? focused
    const recovered = yield* recover(paths, stagingTab, tabId, focusedLive).pipe(Effect.result)

    if (Result.isFailure(recovered)) {
      output.stderr.push(
        `mosaic: recovery failed; panes remain in ${stagingTab}: ${errorText(recovered.failure)}`,
      )
    }
  }

  return yield* applied.failure
})

const runAction = Effect.fnUntraced(function*(
  paths: PluginPathValues,
  output: CapturedOutput,
  action: string,
) {
  const direction = action === "resize-left"
    || action === "resize-down"
    || action === "resize-up"
    || action === "resize-right"
    ? RESIZE_DIRECTIONS[action]
    : undefined

  if (direction !== undefined) {
    const paneId = paneContext(paths)

    if (paneId === undefined) {
      return yield* new LayoutError({ message: "resize action requires a pane context" })
    }

    yield* rpcCall("pane.resize", {
      pane_id: paneId,
      direction,
      amount: RESIZE_AMOUNT,
    })

    return 0
  }

  if (action === "equalize" || action === "cycle") {
    const current = yield* exportLayout(paths, undefined)

    yield* rearrangePanes(paths, output, current, targetFor(action, current.root))

    return 0
  }

  return yield* new LayoutError({ message: `unknown layout action ${pythonRepr(action)}` })
})

const reportLayoutFailure = Effect.fnUntraced(function*(
  paths: PluginPathValues,
  output: CapturedOutput,
  error: LayoutError | RpcTransportError | FlockError,
) {
  const message = errorText(error)

  yield* pluginWarn(paths, output, `layouts: ${message}`)
  yield* notifyFailure(message)
  output.stderr.push(`mosaic: ${message}`)

  return 1
})

export const runLayout = Effect.fnUntraced(function*(argv: readonly string[]) {
  const paths = yield* PluginPaths
  const output = emptyOutput()
  const action = argv[0] ?? ""

  const code = yield* withExclusiveLock(
    pluginLockPath(paths.stateDir),
    runAction(paths, output, action),
  ).pipe(
    Effect.catchTags({
      LayoutError: (error) => reportLayoutFailure(paths, output, error),
      RpcTransportError: (error) => reportLayoutFailure(paths, output, error),
      FlockError: (error) => reportLayoutFailure(paths, output, error),
      LockTimeout: () =>
        pluginWarn(paths, output, "layout: timed out waiting for plugin lock").pipe(Effect.as(1)),
    }),
  )

  return commandResult(code, output)
})
